### PR TITLE
[LTC] Upstream class BackendDevice

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68027

Summary:
This commit upstreams class BackendDevice to the master, which is a backend
specific representation of the actual hardware, for instances, CPU, GPU, or
TPU.

This concept is important for backend like XLA where it needs to tell the
actual hardware type from the c10::DeviceType::Lazy virtual device during
both IR constructions and lowerings.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D32261838](https://our.internmc.facebook.com/intern/diff/D32261838)

### CODE REVIEW

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32261838).

File: torch/csrc/lazy/backend/backend_device.cpp
Comment (alanwaketan) [inline torch/csrc/lazy/backend/backend_device.cpp:16]: Here is the fix, @wconstab.

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32261838).

**Comment by alanwaketan** [comment]
This pull request is ready for review.

File: torch/csrc/lazy/backend/backend_device.h
Comment (wconstab) [inline torch/csrc/lazy/backend/backend_device.h:36]: am I supposed to delete this right away, once I merge my changes for backend interface? or we need to keep it around for a while longer?

File: torch/csrc/lazy/backend/backend_device.h
Comment (wconstab) [inline torch/csrc/lazy/backend/backend_device.h:18]: did you consider tradeoffs of making this a pure-interface class vs having implementation that returns 'Unknown'?

**Review by wconstab — APPROVED** [review]

File: torch/csrc/lazy/backend/backend_device.h
Comment (alanwaketan) [reply torch/csrc/lazy/backend/backend_device.h:36]: Good question. I think you can just delete this ctor during or after your merge PR.

File: torch/csrc/lazy/backend/backend_device.h
Comment (alanwaketan) [reply torch/csrc/lazy/backend/backend_device.h:18]: It's probably just a policy. Do we want to enforce all backends to implement this interface? I think backends should implement this but I don't know if we want to force that. Things can still work even if they don't, for instance, TS backend. Just for UAPI that query the actual hardware type won't work. So, I'm leaning towards to recommended but not required.

Comment (wconstab) [reply torch/csrc/lazy/backend/backend_device.h:18]: yea that makes sense.  I am fine with this.

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32261838).

File: torch/csrc/lazy/backend/backend_device.h
Comment (JackCaoG) [reply torch/csrc/lazy/backend/backend_device.h:36]: yea, backend can probably just add this if they want to(pt/xla will mostly add it on the extended Device class) .

**Review by JackCaoG — APPROVED** [review]
Thanks!

**Comment by alanwaketan** [comment]
Thanks for approving this pull request, @wconstab and @JackCaoG. I will wait for the CI signals before landing.

**Pull request closed without merging**
